### ISSUE TITLE
bitcoin executables segmentation fault upon termination

### ISSUE THREAD

**Issue by ftrader** [gandrewstone/bitcoin#113]
I've noticed this problem with BU 0.12.1bu (built from current branch) on my Debian 7 x86_64 test platform so far.

When I close bitcoin-qt by clicking on the 'X' window decorator, the application shuts down, but sometimes hits a 'Segmentation Fault' in various ways. This does not happen always (maybe 20-30% of the time? - I havent measured accurately).

I've also noticed a similar problem when the bitcoind executable is terminated, e.g. for a regtest is stopped.
Sometimes this segfaulting manifests in backtraces from glibc.

I have yet to try on other platforms (more recent Debian / Ubuntu etc).

@ptschip reported seeing similar crashes when running the wallet.py script.

**Comment by nomnombtc** [comment]
I see also some weirdness on shutdown here (0.12.1bu / f1bb2eb), currently I run with `--enable-debug` because of a different issue. For me it seems to end in SIGABRT, but maybe it is related and helps to find the issue so here is a backtrace:

```
Core was generated by `bitcoind -pid=/var/run/bitcoin/bitcoind.pid -daemon'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007ff6d81b63f8 in raise () from /lib64/libc.so.6
[Current thread is 1 (Thread 0x7ff6dad71740 (LWP 28521))]
(gdb) bt
#0  0x00007ff6d81b63f8 in raise () from /lib64/libc.so.6
#1  0x00007ff6d81b7ffa in abort () from /lib64/libc.so.6
#2  0x00007ff6d81f86f9 in __libc_message () from /lib64/libc.so.6
#3  0x00007ff6d81ff783 in malloc_consolidate () from /lib64/libc.so.6
#4  0x00007ff6d82011e8 in _int_free () from /lib64/libc.so.6
#5  0x00007ff6d8204c1c in free () from /lib64/libc.so.6
#6  0x0000563997ae916c in CNetCleanup::~CNetCleanup (this=0x5639984c7518 <cnet_instance_cleanup>, __in_chrg=<optimized out>) at net.cpp:2280
#7  0x00007ff6d81bafb8 in __run_exit_handlers () from /lib64/libc.so.6
#8  0x00007ff6d81bb005 in exit () from /lib64/libc.so.6
#9  0x00007ff6d81a17d7 in __libc_start_main () from /lib64/libc.so.6
#10 0x00005639978ce999 in _start () at ../sysdeps/x86_64/start.S:108
```

There was also this in debug.log:

```
2016-10-10 05:05:04 Shutdown: done
2016-10-10 05:05:05 CloseSocket(hListenSocket) failed with error Bad file descriptor (9)
2016-10-10 05:05:05 CloseSocket(hListenSocket) failed with error Bad file descriptor (9)
```

**Comment by Justaphf** [comment]
Based on the line number in your backtrace seems like it is complaining about deleting the semaphore for tracking outbound connections, `semOutbound` in the `CNetCleanup` destructor.

Wondering if this is related to the way `StopNode()` calls `semOutbound->post()` on this semaphore based on the number of configured max outbound connections.  I had seen this code before when looking into another issue and thought it looked odd given the way that `post()` doesn't do anything special when in shutdown mode so if there were any threads blocked waiting on this semaphore to release those blocked threads would start executing like normal in the middle of the shutdown routine, possibly adding a new node to the `vNodes` vector via the call to `ConnectNode()` inside of `OpenNetworkConnection()` after (or during) the clean-up of vNodes?

I'm not familiar enough with C++/Boost multi-threading to be sure.  @ptschip does this theory seem plausible?

**Comment by ptschip** [comment]
i tend to agree, i don't see why the post has to been done there...looks
like it might cause trouble as you seem to have found.  I think you can
put in a PR to comment out those line ins stopnode()  where it does that.

**Comment by Justaphf** [comment]
Upon further investigation, I'm not so sure the scenario above is the culprit.  `StopNode()` is only called from inside of `Shutdown()`, which in turn is only called after calls to `Interrupt()` and `join_all()` to ensure all threads have completed/terminated (assuming I'm understanding the thread signaling/interrupts correctly).  There is one case where the `join_all()` is commented out, but that only occurs if you try to shut down before the initialization routine is completed during application startup.

A couple of other things that I saw which stood out, but may not be related are as follows:
- Both `semOutbound` and `semOutboundAddNode` are defined in `unlimited.cpp`, however in `net.cpp` one is declared `extern` and the other `static`.  I believe that `extern` is the correct usage here since they are defined in `unlimited.cpp`.  Technically both of these could be moved wholly into net.cpp since neither are used outside of this file and eliminate the need for either `static` or `extern`.
- The use of `semOutbound` to obtain grants is immediately followed by an `interruption_point()`, however grants on `semOutboundAddNode` don't do this.  Not really sure if this is necessary though since both usages of `semOutboundAddNode` are immediately followed by calls to `OpenNetworkConnection()` which has it's own `interruption_point()` right at the start of the function.

I'm adding some log statements to see if I can verify the order in which various things are being called during shutdown, however running on Windows I'm not seeing errors except for occasionally some silent memory leaks logged in the System Event Logs (RADAR_LEAK_WOW64 and RADAR_PRE_LEAK_WOW64).

**Comment by ptschip** [comment]
if you're trying to re-create the problem I find running the wallet.py
test script will create a segfault on shutdown about 30 to 50% of the
time.   I believe also there is something not quite right in
FinalizeNode or Netcleanup.    I also wonder about the "CNetCleanup
cnet_instance_cleanup_attribute..."  found near the top of
unlimited.cpp.  There is comment there are it causing a segfault in osx
on shutdown, or it did at one time...not sure what it does, something to
do with statistics gathering.

  thatOn 11/10/2016 11:47 PM, Justaphf wrote:

**Comment by Justaphf** [comment]
@ptschip Thanks, I was able to reproduce running the wallet test and the issue seems to be in the clean-up of the local node `delete pnodeLocalHost` when destroying the bloom filter inside of the `~CNode` (see stack trace below).  I haven't been able to trace down why this happens, but commenting out the clean-up code for this pointer, while causing a memory leak, eliminates the segfault.

If you want I can put in a band-aid PR for this, but I'd rather figure out the root cause instead of avoiding the segfault by allowing a memory leak.

```
(gdb) bt
#0  malloc_consolidate (av=av@entry=0x7f2a17b83620 <main_arena>) at malloc.c:4149
#1  0x00007f2a17857641 in _int_free (av=0x7f2a17b83620 <main_arena>, p=<optimized out>, have_lock=0) at malloc.c:4057
#2  0x00007f2a1a28201d in deallocate (this=<optimized out>, __p=<optimized out>)
    at /usr/include/c++/4.9/ext/new_allocator.h:110
#3  deallocate (__a=..., __n=<optimized out>, __p=<optimized out>) at /usr/include/c++/4.9/ext/alloc_traits.h:185
#4  _M_deallocate (this=<optimized out>, __n=<optimized out>, __p=<optimized out>)
    at /usr/include/c++/4.9/bits/stl_vector.h:178
#5  ~_Vector_base (this=0x7f2a1c3681e8, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/stl_vector.h:160
#6  ~vector (this=0x7f2a1c3681e8, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/stl_vector.h:425
#7  ~CBloomFilter (this=0x7f2a1c3681e8, __in_chrg=<optimized out>) at bloom.h:45
#8  ~CRollingBloomFilter (this=0x7f2a1c3681b8, __in_chrg=<optimized out>) at bloom.h:117
#9  CNode::~CNode (this=this@entry=0x7f2a1c367d60, __in_chrg=<optimized out>) at net.cpp:2664
#10 0x00007f2a1a282623 in CNetCleanup::~CNetCleanup (this=<optimized out>, __in_chrg=<optimized out>) at net.cpp:2285
#11 0x00007f2a17815b29 in __run_exit_handlers (status=0, listp=0x7f2a17b835a8 <__exit_funcs>,
    run_list_atexit=run_list_atexit@entry=true) at exit.c:82
#12 0x00007f2a17815b75 in __GI_exit (status=<optimized out>) at exit.c:104
#13 0x00007f2a177ffb4c in __libc_start_main (main=0x7f2a1a199470 <main(int, char**)>, argc=8, argv=0x7ffc76e19aa8,
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffc76e19a98) at libc-start.c:321
#14 0x00007f2a1a1a5b71 in _start ()
```

**Comment by ptschip** [comment]
that's great!
and yes i'd rather figure out the root cause...at least it sounds like
your getting close.

**Comment by ptschip** [comment]
there is no lock on the filter before deleting it...i think it's
possible that pfilter is getting updated at the same time we're doing
cleanup. you might want to try that.

**Comment by Justaphf** [comment]
Thanks for the tip @ptschip.  Unfortunately I have a pending customer release at work which has been monopolizing my time and I doubt I'll be able to look into this further for at least another week or two.  If someone wants to take over looking into this, please do.  Otherwise I'll look into it once my work hours return to normal.

Anyone who wants to take over please note that the `CNode` actually has instances of 6 `CBloomFilter` objects in the BU implementation and I'm not sure which is causing the issue.  There are two direct references in the form of `pfilter` and `pThinBlockFilter`, which are initialized and maintained via pointers, and there are 4 indirect references in the form of 2 `CRollingBloomFilter` reference instances which contain 2 `CBloomFilter` reference instances each for `addrKnown` and `filterInventoryKnown`.  Based on the stack trace that I posted above, the issue seems to be occurring with the destruction of one of the `CRollingBloomFilter` objects.  None of these objects have synchronization objects associated with them so  as ptschip suggested, adding in multi-thread sync is probably the fix.

**Comment by sickpig** [comment]
@ftrader @nomnombtc could please test @Justaphf #128

**Comment by ftrader** [comment]
@sickpig : the PR#128 fix eliminates the segfaults on exit on my Linux platforms, as far as preliminary testing indicates.

**Comment by sickpig** [comment]
I'm not able to reproduce this anymore on current 0.12.1bu, maybe due to fact that PR #138 and #140 have been merged. 
@ftrader are you able to reproduce the issue? if it's not the case I think we could close this issue.

**Comment by nomnombtc** [comment]
@sickpig also tested it again with current 0.12.1bu bacc271.

I tried the `wallet.py` test because this was causing segfaults on node shutdown earlier:
**Debian Jessie, Debian Stretch, OpenSUSE Leap 42.1, Fedora 24, Archlinux - all work fine without any problem.**

But on slackware I still get segfaults in `dmesg`, super strange and it makes no sense to me. Tried multiple versions 13.37, 14.1, 14.2 all the same. I just don't understand what could cause such behavior, if anyone has an idea...

To see if it does the same I tried with bitcoin-core 0.12 branch, but it does not segfault when running `wallet.py` test, everything good. So slackware can't be horribly broken and for me this still is some kind of very strange regression.

But it has improved a lot, `bitcoin-qt` was segfaulting for me at every shutdown this does not happen anymore, now it only happens when I run `wallet.py` test.

Slackware backtrace (edit: better bt from a fresh installed slackware64-14.2 vm):

```
#0  0x00007ff0d735c578 in malloc_consolidate () from /lib64/libc.so.6
#1  0x00007ff0d735e1e8 in _int_free () from /lib64/libc.so.6
#2  0x00007ff0d7361c1c in free () from /lib64/libc.so.6
#3  0x000055c73a389c31 in __gnu_cxx::new_allocator<unsigned char>::deallocate (this=0x55c73d608348, __p=<optimized out>) at /usr/include/c++/5.3.0/ext/new_allocator.h:110
#4  __gnu_cxx::__alloc_traits<std::allocator<unsigned char> >::deallocate (__a=..., __n=<optimized out>, __p=<optimized out>) at /usr/include/c++/5.3.0/ext/alloc_traits.h:185
#5  std::_Vector_base<unsigned char, std::allocator<unsigned char> >::_M_deallocate (__n=<optimized out>, __p=<optimized out>, this=0x55c73d608348) at /usr/include/c++/5.3.0/bits/stl_vector.h:178
#6  std::_Vector_base<unsigned char, std::allocator<unsigned char> >::~_Vector_base (this=0x55c73d608348, __in_chrg=<optimized out>) at /usr/include/c++/5.3.0/bits/stl_vector.h:160
#7  std::vector<unsigned char, std::allocator<unsigned char> >::~vector (this=0x55c73d608348, __in_chrg=<optimized out>) at /usr/include/c++/5.3.0/bits/stl_vector.h:425
#8  CBloomFilter::~CBloomFilter (this=0x55c73d608348, __in_chrg=<optimized out>) at bloom.h:45
#9  CRollingBloomFilter::~CRollingBloomFilter (this=0x55c73d608318, __in_chrg=<optimized out>) at bloom.h:117
#10 CNode::~CNode (this=0x55c73d607ec0, __in_chrg=<optimized out>) at net.cpp:2702
#11 0x000055c73a38a24e in CNetCleanup::~CNetCleanup (this=<optimized out>, __in_chrg=<optimized out>) at net.cpp:2323
#12 0x00007ff0d7317fb8 in __run_exit_handlers () from /lib64/libc.so.6
#13 0x00007ff0d7318005 in exit () from /lib64/libc.so.6
#14 0x00007ff0d72fe7d7 in __libc_start_main () from /lib64/libc.so.6
#15 0x000055c73a2aa329 in _start () at ../sysdeps/x86_64/start.S:108
```

**Comment by nomnombtc** [comment]
@sickpig @gandrewstone I tried it again here with  0.12.1bu 8828196 and I can report that it works fine now on slackware, no core dumps anymore :+1:

**Comment by ftrader** [comment]
I don't get anymore of these segfaults now either.

Going to close the issue now. Thanks everyone!

**Comment by sickpig** [comment]
@ftrader thanks a lot for your persistence and effort! 
great work @gandrewstone at fixing it!
